presto-hive/src/main/java/com/facebook/presto/hive/util/Statistics.java

```diff
@@ -59,22 +94,72 @@ public static HiveBasicStatistics reduce(HiveBasicStatistics first, HiveBasicSta
     {
-        return type.equals(TINYINT)
-                || type.equals(SMALLINT)
-                || type.equals(INTEGER)
-                || type.equals(BIGINT)
-                || type.equals(REAL)
-                || type.equals(DOUBLE)
-                || type.equals(DATE)
-                || type.equals(TIMESTAMP)
-                || (type instanceof DecimalType);
+        // skip the columns for which statistics from either of sides are missing
+        Set<String> columns = intersection(first.keySet(), second.keySet());
+        return columns.stream()
+                .collect(toImmutableMap(column -> column, column -> merge(first.get(column), firstRowCount, second.get(column), secondRowCount)));
+    }
+
+    public static HiveColumnStatistics merge(HiveColumnStatistics first, OptionalLong firstRowCount, HiveColumnStatistics second, OptionalLong secondRowCount)
+    {
+        return new HiveColumnStatistics(
+                first.getLowValue().isPresent() ? reduce(first.getLowValue(), second.getLowValue(), SELECT_MIN) : second.getLowValue(),
+                first.getHighValue().isPresent() ? reduce(first.getHighValue(), second.getHighValue(), SELECT_MAX) : second.getHighValue(),
+                reduce(first.getMaxColumnLength(), second.getMaxColumnLength(), SELECT_MAX),
+                mergeAvg(first.getAverageColumnLength(), firstRowCount, second.getAverageColumnLength(), secondRowCount),
+                reduce(first.getTrueCount(), second.getTrueCount(), ADD),
+                reduce(first.getFalseCount(), second.getFalseCount(), ADD),
+                reduce(first.getNullsCount(), second.getNullsCount(), ADD),
+                reduce(first.getDistinctValuesCount(), second.getDistinctValuesCount(), SELECT_MAX));
+    }
+
+    private static OptionalDouble mergeAvg(OptionalDouble first, OptionalLong firstRowCount, OptionalDouble second, OptionalLong secondRowCount)
+    {
+        if (first.isPresent() && second.isPresent() && firstRowCount.isPresent() && secondRowCount.isPresent()) {
+            double sumFirst = first.getAsDouble() * firstRowCount.getAsLong();
+            double sumSecond = second.getAsDouble() * secondRowCount.getAsLong();
+            long totalRowCount = firstRowCount.getAsLong() + secondRowCount.getAsLong();
+            return OptionalDouble.of((sumFirst + sumSecond) / totalRowCount);
+        }
+        return OptionalDouble.empty();
+    }
+
+    public static Set<ColumnStatisticType> getSupportedStatistics(Type type)
+    {
+        if (type.equals(BOOLEAN)) {
+            return ImmutableSet.of(NUMBER_OF_NON_NULL_VALUES, NUMBER_OF_TRUE);
+        }
+        else if (type.equals(BIGINT) || type.equals(INTEGER) || type.equals(SMALLINT) || type.equals(TINYINT)) {
+            return ImmutableSet.of(NUMBER_OF_NON_NULL_VALUES, MIN, MAX, NUMBER_OF_DISTINCT_VALUES);
+        }
+        else if (type.equals(DOUBLE) || type.equals(REAL)) {
```

Comment:
Let's combine the branches all of these numeric-ish types (including data/timestamp), so it's clear we are collecting the same stats

Reply:
Changed to `if (isNumericType(type) || type.equals(DATE) || type.equals(TIMESTAMP)) {`